## week 2-3 preprocessing homework
#### social media interactions dataset preprocessing notebook

digma, ronachel marie of h3101

### import libraries and dataset

In [2]:
import pandas as pd
import numpy as np

# load dataset

interactions = pd.read_csv('../data/raw/social_media_interactions_contaminated.csv')

### inspect

In [3]:
# inspect dataset
print("First 20 rows: ")
display(interactions.head(20))

print("\nDataset Information: ")
display(interactions.info())

print("\nDataset Dimensions: ")
display(interactions.shape)

print("\nDataset Columns: ")
display(interactions.columns)

print("\nDataset Numerical Statistics: ")
display(interactions.describe())

First 20 rows: 


,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
0,2dcb9523-356b-40b2-a67b-1f27797de261,e5d15761-d0a7-4329-89e3-79a892c56097,2023-07-11,NaN,Comment,NaN
1,e12c37b3-7d4d-472f-9fd8-0df2cb3001aa,02f9f376-70ae-4fcd-9070-1db977939948,2023-07-06,Twitter,Share,NaN
2,08a911a3-65e6-4f5d-a6a1-ae7ddcbe28a2,a83fa04c-f109-4f24-8ce1-2078154f6a1c,2024-05-24,Instagram,Comment,Neutral
3,efdfdfc9-5dbb-4478-911a-101a390a0285,28a69c4b-a2e4-4c74-a130-1132d7733fdf,2023-11-01,Instagram,Like,Neutral
4,ca1e90f6-0e5f-492e-ab92-252ff540da18,d9d1c6f8-5e15-4738-b52b-13c2982420cc,2023-07-08,Instagram,Like,NaN
5,3e44871b-f56c-4576-b1ca-d1dc999e2166,0c409883-8396-48e4-83fb-887329848696,2023-12-18,Instagram,Comment,Positive
6,aa5eea4b-c948-41f4-9285-229a470002aa,4034dadf-6541-40d6-a7f0-16b20a009c04,2023-11-15,Instagram,Share,Positive
7,7d83304d-2501-4c9f-ba63-a1a14343e51f,bcfc43e7-c5aa-4dbd-9961-0e108784b199,2024-03-29,Instagram,Comment,NaN
8,bd90f5cb-05a7-40f4-acb1-eedf48b58ffa,d2a06664-703d-4bc1-9401-a06f8c43fda5,2024-05-02,NaN,Like,Neutral
9,63ee220c-19ae-4113-b45c-276b22b068e1,67cdfb0b-3da6-46b8-a1f4-03b2abc81f58,2023-11-18,NaN,Comment,Positive



Dataset Information: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3200 entries, 0 to 3199
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   CustomerID       3200 non-null   object
 1   InteractionID    3200 non-null   object
 2   InteractionDate  3200 non-null   object
 3   Platform         2889 non-null   object
 4   InteractionType  3200 non-null   object
 5   Sentiment        2871 non-null   object
dtypes: object(6)
memory usage: 150.1+ KB


None


Dataset Dimensions: 


(3200, 6)


Dataset Columns: 


Index(['CustomerID', 'InteractionID', 'InteractionDate', 'Platform',
       'InteractionType', 'Sentiment'],
      dtype='object')


Dataset Numerical Statistics: 


,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
count,3200,3200,3200,2889,3200,2871
unique,1893,3000,450,3,3,5
top,fd35b6b4-ad1c-49cc-9c34-8c2d696f38e2,e77a8713-fdc7-4f71-85f8-9cd88cca0c42,2023-07-02,Instagram,Comment,Positive
freq,6,2,17,976,1079,962


### exact duplicates

In [4]:
# inspect duplicates
duplicate_interactions = interactions.duplicated().sum()

print("Exact duplicate rows:", duplicate_interactions)

display(interactions[interactions.duplicated(keep=False)])

Exact duplicate rows: 180


,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
3,efdfdfc9-5dbb-4478-911a-101a390a0285,28a69c4b-a2e4-4c74-a130-1132d7733fdf,2023-11-01,Instagram,Like,Neutral
30,eb8bb3ce-e182-4a9c-8084-c55c738815e1,ee2c3d14-248c-48b0-aa80-4ea97e449bf8,2023-12-22,Instagram,Comment,Negative
64,58ced86b-155c-45b1-96c4-92e39af3751e,7d7f1f20-150c-491c-9dc3-fe2922a19836,2023-10-10,Instagram,Comment,Negative
71,a7b667f7-d6c7-4fc6-b6d9-bcdeea32297c,39622755-9e00-4107-8ee1-5defd5a5dbab,2023-10-06,Twitter,Like,Neutral
91,47bf3bc6-17a6-4b21-82b5-90f78af4c25c,4eac4bf7-b0ff-4eb9-af31-69b5413a99af,2023-10-19,Instagram,Like,Positive
...,...,...,...,...,...,...
3195,8d8fe8bd-2009-4a15-be37-83dba3cef447,2212803c-4a00-4263-bae5-aaaff6d6bdb1,2023-08-30,Instagram,Like,Neutral
3196,1756838b-8368-4852-b4a5-32a23643a3dd,2f50ca57-f930-4d3e-ae08-9f3b26c7f148,2024-06-16,Facebook,Like,Neutral
3197,eb8bb3ce-e182-4a9c-8084-c55c738815e1,ee2c3d14-248c-48b0-aa80-4ea97e449bf8,2023-12-22,Instagram,Comment,Negative
3198,a5d08a95-9dba-4e3d-b997-1cfdde6a89ef,e77a8713-fdc7-4f71-85f8-9cd88cca0c42,2023-11-01,Twitter,Share,Positive


In [8]:
# drop exact duplicate records
interactions_cleaned = interactions.drop_duplicates().copy()

# verify and check shape 
print("Duplicates remaining:", interactions_cleaned.duplicated().sum())
print("Dataset shape after removal of duplicates:", interactions_cleaned.shape)

Duplicates remaining: 0
Dataset shape after removal of duplicates: (3020, 6)


removing exact duplicate records

the dataset initially contained 180 exact duplicate rows. these records contain identical values across all columns and therefore do not provide additional information for analysis so these exact duplicate rows were removed to prevent duplicated interactions from disproportionately influencing subsequent analysis and machine learning processes.

### missing, invalid, and inconsistent values

#### interaction id

In [2]:
# check duplicate interaction ids
print("Duplicate InteractionIDs:", interactions_cleaned["InteractionID"].duplicated().sum())

# check number of records with unique interaction ids
print("\nNumber of Unique InteractionIDs:", interactions_cleaned["InteractionID"].nunique())

NameError: name 'interactions_cleaned' is not defined

checking uniqueness of interaction ids

after removing exact duplicate rows, 20 `InteractionID` values were still duplicated. these records were inspected to determine whether they represented exact duplicates or potentially distinct interactions.

the duplicated records were found to contain differences in fields such as `InteractionDate` and `Sentiment`. since there was insufficient information to determine which record was correct, the records were retained rather than arbitrarily deleting one. these duplicated IDs are documented as data-quality anomalies.

In [ ]:
# verify
duplicate_interaction_ids = interactions_cleaned[interactions_cleaned.duplicated(subset="InteractionID", keep=False)].sort_values("InteractionID")

print("Records involved in duplicated InteractionIDs:", len(duplicate_interaction_ids))
display(duplicate_interaction_ids)

Records involved in duplicated InteractionIDs: 40


,CustomerID,InteractionID,InteractionDate,Platform,InteractionType,Sentiment
374,11b046f0-b091-4bfb-b0f4-df58bfea039a,01f1c8c8-df21-4d75-adbb-feb6d4c84317,29/04/2024,Instagram,Like,Positive
3005,11b046f0-b091-4bfb-b0f4-df58bfea039a,01f1c8c8-df21-4d75-adbb-feb6d4c84317,2024-03-20,Instagram,Like,Positive
3093,2ddc8052-fbbe-471c-a43e-54b81467423d,04391eb2-8391-44d8-9f45-29edcde96e5d,2023-11-30,Twitter,Share,Very Positive
2183,2ddc8052-fbbe-471c-a43e-54b81467423d,04391eb2-8391-44d8-9f45-29edcde96e5d,2023-11-30,Twitter,Share,Negative
2532,e06f0e16-8e48-483b-b103-9f105b139ca8,071a2657-12a8-4216-9c64-baf9ab5c9420,2024-04-30,Twitter,Share,Very Positive
3166,e06f0e16-8e48-483b-b103-9f105b139ca8,071a2657-12a8-4216-9c64-baf9ab5c9420,2024-04-30,Twitter,Share,Neutral
3084,68afa4cc-e2d5-4150-b358-2e14e7508721,0942889b-a3a9-4039-b4b0-12fb8714c95f,24/01/2024,Twitter,Comment,Negative
2490,68afa4cc-e2d5-4150-b358-2e14e7508721,0942889b-a3a9-4039-b4b0-12fb8714c95f,2024-06-14,Twitter,Comment,Negative
3017,a44b6ca5-c73e-4f7a-919c-617931e33a29,13b59881-5aaf-4510-9dcf-4f9d116bf645,04/06/2024,Instagram,Like,Positive
2135,a44b6ca5-c73e-4f7a-919c-617931e33a29,13b59881-5aaf-4510-9dcf-4f9d116bf645,2023-07-28,Instagram,Like,Positive


#### platform

In [ ]:
# since there were 311 missing values in the platform column that came out during initial inspection, inspect the partially cleaned dataset
print("Platform values:", interactions_cleaned["Platform"].value_counts(dropna=False))
print("\nUnique Platform values:", interactions_cleaned["Platform"].dropna().unique())

Platform values: Platform
Instagram    921
Twitter      917
Facebook     891
NaN          291
Name: count, dtype: int64

Unique Platform values: ['Twitter' 'Instagram' 'Facebook']


missing platform values

the platform olumn contains three consistently formatted categories: 'Twitter' 'Instagram' 'Facebook', no inconsistent platform labels were identified.

after removing exact duplicate records, 291 platform column values remained missing. since the correct platform cannot be reliably determined from the available data, the missing values were replaced with unknown rather than assigning an unsupported platform or removing the interaction records. this preserves the observations while explicitly identifying the missing information.

In [ ]:
# fill missing platform values with 'unknown'
interactions_cleaned["Platform"] = interactions_cleaned["Platform"].fillna("Unknown")

# verify
print("Missing Platform values:", interactions_cleaned["Platform"].isna().sum())
print("\nPlatform values:", interactions_cleaned["Platform"].value_counts())

Missing Platform values: 0

Platform values: Platform
Instagram    921
Twitter      917
Facebook     891
Unknown      291
Name: count, dtype: int64


#### sentiment

In [ ]:
# inspect column
print("Sentiment values:", interactions_cleaned["Sentiment"].value_counts(dropna=False))
print("\nUnique Sentiment values:", interactions_cleaned["Sentiment"].dropna().unique())

Sentiment values: Sentiment
Positive         905
Neutral          874
Negative         866
NaN              309
Very Negative     34
Very Positive     32
Name: count, dtype: int64

Unique Sentiment values: ['Neutral' 'Positive' 'Negative' 'Very Negative' 'Very Positive']


missing sentiment values
the sentiment column contains 5 consistently formatted categories: 'Neutral' 'Positive' 'Negative' 'Very Negative' 'Very Positive'. no inconsistent labels were identified among the existing values.

after removing exact duplicate records, 309 sentiment values remained missing. since sentiment cannot be reliably inferred from the available fields, the missing values were replaced with 'unknown' rather than assigning an unsupported sentiment category or removing the interaction records.

In [ ]:
# fill missing sentiment values with 'unknown'
interactions_cleaned["Sentiment"] = interactions_cleaned["Sentiment"].fillna('Unknown')

# verify
print("Missing Sentiment values:", interactions_cleaned["Sentiment"].isna().sum())
print("\nSentiment values:", interactions_cleaned["Sentiment"].value_counts())

Missing Sentiment values: 0

Sentiment values: Sentiment
Positive         905
Neutral          874
Negative         866
Unknown          309
Very Negative     34
Very Positive     32
Name: count, dtype: int64


#### interaction date

In [ ]:
# inspect interaction date formats
print("InteractionDate data type:", interactions_cleaned["InteractionDate"].dtype)

print("\nSample InteractionDate values:")
print(interactions_cleaned["InteractionDate"].head(20).to_string(index=False))

print("\nNumber of unique date values:", interactions_cleaned["InteractionDate"].nunique())

print("\nExamples containing '/':")
print(interactions_cleaned[interactions_cleaned["InteractionDate"].astype(str).str.contains("/", regex=False)]["InteractionDate"].head(20).to_string(index=False))

InteractionDate data type: object

Sample InteractionDate values:
2023-07-11
2023-07-06
2024-05-24
2023-11-01
2023-07-08
2023-12-18
2023-11-15
2024-03-29
2024-05-02
2023-11-18
2024-05-28
2024-06-27
2023-08-31
2023-08-27
2024-06-10
2024-02-12
2023-08-18
2023-08-17
2024-03-28
2023-10-08

Number of unique date values: 450

Examples containing '/':
26/01/2024
01/10/2023
28/03/2024
15/08/2023
01/01/2024
04/04/2024
12/08/2023
01/09/2023
05/09/2023
01/03/2024
03/05/2024
27/10/2023
05/01/2024
12/03/2024
01/02/2024
29/04/2024
21/05/2024
21/09/2023
06/07/2023
21/08/2023


standardize interaction dates

the interaction date column was initially stored as a string and contained dates in multiple formats, including `yyyy-mm-dd` and `dd/mm/yyyy`.

the column was converted to a datetime data type using mixed-format parsing with `dayfirst=True`. invalid values were coerced to missing values so they could be identified during validation.

In [ ]:
# convert data type
interactions_cleaned["InteractionDate"] = pd.to_datetime(interactions_cleaned["InteractionDate"], format="mixed", dayfirst=True, errors="coerce")

# verify successful conversion
print("InteractionDate data type:", interactions_cleaned["InteractionDate"].dtype)

print("Missing InteractionDate values:", interactions_cleaned["InteractionDate"].isna().sum())

print("\nDate range:")
print("Earliest:", interactions_cleaned["InteractionDate"].min())
print("Latest:", interactions_cleaned["InteractionDate"].max())

InteractionDate data type: datetime64[ns]
Missing InteractionDate values: 0

Date range:
Earliest: 2023-07-01 00:00:00
Latest: 2024-06-30 00:00:00


the interaction date column was successfully converted from an object/string data type to datetime64[ns]. the mixed date formats were parsed correctly using day-first interpretation for slash-formatted dates.

validation confirmed that there are no missing or invalid interaction dates, and the resulting interaction dates range from July 1, 2023 to June 30, 2024.

### final validation

In [30]:
print("FINAL SOCIAL MEDIA DATASET VALIDATION")
print("=" * 50)

print("\nDataset Shape:")
print(interactions_cleaned.shape)

print("\nData Types:")
print(interactions_cleaned.dtypes)

print("\nMissing Values:")
print(interactions_cleaned.isna().sum())

print("\nExact Duplicate Rows:")
print(interactions_cleaned.duplicated().sum())

print("\nDuplicate InteractionIDs:")
print(interactions_cleaned["InteractionID"].duplicated().sum())

print("\nPlatform Values:")
print(interactions_cleaned["Platform"].value_counts())

print("\nSentiment Values:")
print(interactions_cleaned["Sentiment"].value_counts())

print("\nInteractionDate Validation:")
print("Date data type:", interactions_cleaned["InteractionDate"].dtype)
print("Earliest date:", interactions_cleaned["InteractionDate"].min())
print("Latest date:", interactions_cleaned["InteractionDate"].max())

FINAL SOCIAL MEDIA DATASET VALIDATION

Dataset Shape:
(3020, 6)

Data Types:
CustomerID                 object
InteractionID              object
InteractionDate    datetime64[ns]
Platform                   object
InteractionType            object
Sentiment                  object
dtype: object

Missing Values:
CustomerID         0
InteractionID      0
InteractionDate    0
Platform           0
InteractionType    0
Sentiment          0
dtype: int64

Exact Duplicate Rows:
0

Duplicate InteractionIDs:
20

Platform Values:
Platform
Instagram    921
Twitter      917
Facebook     891
Unknown      291
Name: count, dtype: int64

Sentiment Values:
Sentiment
Positive         905
Neutral          874
Negative         866
Unknown          309
Very Negative     34
Very Positive     32
Name: count, dtype: int64

InteractionDate Validation:
Date data type: datetime64[ns]
Earliest date: 2023-07-01 00:00:00
Latest date: 2024-06-30 00:00:00


#### summary

after preprocessing, the social media interactions dataset contains 3,020 records and 6 columns. all missing values have been addressed, and no exact duplicate rows remain.

the platform and sentiment columns were checked for consistency. missing platform and sentiment values were labeled as 'unknown' because their correct values could not be reliably inferred from the available data.

The interaction date column was standardized to datetime64[ns], with no invalid or missing dates remaining.

the 20 duplicated interaction id values were investigated and retained because their associated records were not exact duplicates. these records are documented as data-quality anomalies rather than being arbitrarily removed.

### save

In [31]:
interactions_cleaned.to_csv("../data/processed/social_media_interactions_preprocessed.csv",index=False)